Implement basic one-off language inections

Cottand opened the pull request:
Implement `PsiLanguageInjectionHost` in strings, so that they may be injected with the one-off action "Inject Language Or Reference".

There are several not-yet-implemented things, some more important than others. In no particular order:
- `NixStringLiteralEscaper.getOffsetInHost()` is not implemented, because it's hard. This means the action "Edit <lang> Fragment" does not work (it opens a separate window to edit the guest language)
- Indentation when decoding Nix indented string is not implemented. I _think_ this might simply not be possible, because with my attempts editor just crashed. It might be related to how the IDEA API works with "ranges" (the host and the guest language's text ranges) and that might not play well with indented strings, which outright remove a good chunk of text in the middle of the range. The example I was following does not have this problem, because multiline strings are not indented
- Interpolation when decoding Nix intented strings is not escaped. I think this is the least confusing thing to do, but it makes the highlighting UX not ideal when dealing with injected languages that have braces (like JSON).
- Permanent injections are not implemented yet. I am happy to try to do that in a separate PR, I think one-off injections is already an improvement

Another potential issue is that I implemented much of the boilerplate in Kotlin, because I am more at ease writing in it. It's not super in line with the rest of the code, so I am happy to re-write in Java now that the actual development is done, if you prefer.

For reference, I used the [Terraform HCL plugin](https://github.com/JetBrains/intellij-plugins/tree/master/terraform) as my guide, because it also has multiline strings with interpolation.

I am happy to keep working on this PR to implement the things above, or do separate PRs so they are smaller and easier to review. See some pictures below.

Some languages:

<img width="448" alt="image" src="https://github.com/Cottand/nix-idea/assets/45274424/c1bac99a-4249-4fae-bf51-c0d1c69e4bc5">

A warning hint of Yaml inside a Nix string:

<img width="428" alt="image" src="https://github.com/Cottand/nix-idea/assets/45274424/8854bd38-56d0-48e7-989e-b664355ace24">

The fragment editor to see the isolated guest language (this does **not** work if you edit the fragment, because `getOffsetInHost()` is not implemented):

<img width="428" alt="image" src="https://github.com/Cottand/nix-idea/assets/45274424/dc678b05-72ec-48a1-a62e-5a8784c16235">

One line injection:

<img width="399" alt="image" src="https://github.com/Cottand/nix-idea/assets/45274424/890e0670-c7f8-4a3e-953e-5d0ebbac620a">

Cottand commented:
Not sure why it is not passing in CI, `gradle check` passes on my machine :c

JojOatXGME commented:
> Not sure why it is not passing in CI, `gradle check` passes on my machine :c

```
EdtExtensionTest > later_exceptions_are_added_as_suppressed_exceptions() FAILED
    java.lang.AssertionError at EdtExtensionTest.java:78
```

Doesn't look related to your changes. I just triggered a re-build and it succeeded.

JojOatXGME reviewed (changes requested):
Hi, big thanks for your contribution. Unfortunatly, I am abit distracted this week. I may not always respond as fast I would like. Just that you know. Anyway, here is my initial quick review.

> Indentation when decoding Nix indented string is not implemented. I think this might simply not be possible, because with my attempts editor just crashed.

I think IDEA supports that for text blocks in Java. However, that is a rather new think. If I remember correctly, I think IDEA doesn't yet support that for Kotlin. I could imagine that they use a different (maybe internal) API in case of Java. I don't know. I didn't yet really look at the API you are using.

> Another potential issue is that I implemented much of the boilerplate in Kotlin, because I am more at ease writing in it. It's not super in line with the rest of the code, so I am happy to re-write in Java now that the actual development is done, if you prefer.

I don't think that is necessary.

> I am happy to keep working on this PR to implement the things above, or do separate PRs so they are smaller and easier to review. See some pictures below.

Not sure. I am happy to integrate this PR as long as it doesn't break anything.

Cottand commented on src/main/java/org/nixos/idea/psi/impl/AbstractNixPsiElement.java:
```diff
@@ -5,7 +5,7 @@
 import org.nixos.idea.psi.NixPsiElement;
 
-abstract class AbstractNixPsiElement extends ASTWrapperPsiElement implements NixPsiElement {
+abstract public class AbstractNixPsiElement extends ASTWrapperPsiElement implements NixPsiElement {
```
Because `AbstractNixString` needs to extend this class, and it has to be public (because Kotlin does not have package-private), and A cannot extend B if A is public and B is package-private

Writing `AbstractNixString` in Java would solve this

JojOatXGME replied:
Thanks. You don't have to change it. I was just wondering.

Cottand commented on src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt:
```diff
@@ -0,0 +1,33 @@
+
+class NixStringLiteralEscaper(host: AbstractNixString) : LiteralTextEscaper<PsiLanguageInjectionHost>(host) {
+
+    override fun isOneLine(): Boolean = false
```
I took the example from [HCL](https://github.com/JetBrains/intellij-plugins/blob/fb9094045d7492228087a091f2ec248f77825dd8/terraform/src/org/intellij/terraform/hcl/psi/impl/HCLStringLiteralTextEscaper.kt#L12) which also always returns false. I will try to implement proper smart logic, see if that makes any difference, but I suspected it might be best to copy my reference for now 🤷

Cottand commented on src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt:
```diff
@@ -0,0 +1,33 @@
+        if (myHost is NixIndString) {
+            outChars.append(NixIndStringUtil.escape(subText))
+        } else {
+            NixStringUtil.escape(outChars, subText)
```
Ah I misnamed `NixIndStringUtil.escape()`, it should be unescape. I'll fix that, good catch.

JojOatXGME replied:
In this case, I would suggest using `NixStringUtil.parse`, as it already supports both types of strings. You would have to iterate over `NixString.getStringParts()`. Each part implements either `NixStringText` or `NixAntiquotation`. The first one can be unescaped by `NixStringUtil.parse`. The second one represents a string interpolation, not sure what to do with it.

Cottand commented on src/main/java/org/nixos/idea/psi/NixStringManipulator.kt:
```diff
@@ -0,0 +1,37 @@
+import org.nixos.idea.util.NixIndStringUtil
+import org.nixos.idea.util.NixStringUtil
+
+class NixStringManipulator : AbstractElementManipulator<NixString>() {
```
I do not know no, again I followed the example

Your guess is as good as mine! 

The name of `handleContentChange()` makes me suspect this might be called when you chage the text inside the editor fragment of the **guest** language (and this function changes the corresponding string AST node in the **host** language)

Cottand reviewed (commented):
Thanks for your thorough review @JojOatXGME !

> I may not always respond as fast I would like

No hurry at all! I will tag you for review when I think it is ready again but take as long as you need to get back to me. Your reviews are thorough so it must be time consuming. I will be offline next week myself.

> I didn't yet really look at the API you are using.

I am using the Java API so I don't think the problem is there. I suspect dealing with the offsets needs special treatment when passing the decoded text back to the editor (otherwise you might be giving it different ranges it expects?). I am not fully clear on the inner workings of those yet. I will do a bit more exploration, but I suspect I need to implement `decode()` while manually keeping track of offsets like [here](https://github.com/JetBrains/intellij-plugins/blob/master/terraform/src/org/intellij/terraform/hcl/psi/impl/HCLStringLiteralTextEscaper.kt#L35)

JojOatXGME commented:
Hi, looked a bit through the source code of IntelliJ Community to understand how they are handling string interpolations. I noticed the following;

1. They implement `PsiLanguageInjectionHost` in fragments, not the for whole string. In our case, this would mean we would implement `PsiLanguageInjectionHost` in `NixStringText`, instead of `NixString`
2. They implement a `com.intellij.lang.injection.general.LanguageInjectionPerformer` (see [`JavaInjectionPerformer` in IntelliJ source code](https://github.com/JetBrains/intellij-community/blob/7cb4c5a431c611f6cd5caf6a0114384e1edac948/plugins/IntelliLang/java-support/org/intellij/plugins/intelliLang/inject/java/JavaInjectionPerformer.kt))
3. They implement a `com.intellij.lang.injection.MultiHostInjector` (see [`JavaConcatenationToInjectorAdapter` in IntelliJ source code](https://github.com/JetBrains/intellij-community/blob/7cb4c5a431c611f6cd5caf6a0114384e1edac948/java/java-impl/src/com/intellij/psi/impl/source/tree/injected/JavaConcatenationToInjectorAdapter.java))
4. Their `JavaConcatenationToInjectorAdapter` extends `ConcatenationInjectorManager.BaseConcatenation2InjectorAdapter`, which is not part of IDEAs API. `ConcatenationInjectorManager` also provides an extension point using the interface `com.intellij.lang.injection.ConcatenationAwareInjector`. This interface is also again implemented for Java as [`ConcatenationInjector`](https://github.com/JetBrains/intellij-community/blob/7cb4c5a431c611f6cd5caf6a0114384e1edac948/plugins/IntelliLang/java-support/org/intellij/plugins/intelliLang/inject/java/ConcatenationInjector.java).

So, the implementation for Injections in Java seem rather complex, but I didn't yet understand for what all this complexity is for. I hope there is a simpler solution to properly support strings containing interpolations. Anyway, I wanted to share with you what I noticed, although I hoped to provide more specific guidance. I haven't really looked at your reference (Terraform HCL plugin), if they also have injections, maybe there implementation is easier to understand.

Cottand commented:
Hello! Sorry for the delay.

Thanks for taking a look at the Java source. 

> They implement PsiLanguageInjectionHost in fragments, not the for whole string.

I think they do this to be able to perforn injections in concatenated strings, like in `"some: "+"{yaml: stuff}"`. I am not against implementing this for Nix too, but it does make the code more complex, so maybe we can keep it simple for now and revisit in a separate PR if necessary.

I also think it adds complexity, because before when decoding the string (ie, telling IDEA "this is what the string is supposed to be in YAML") you need to concatenate the fragments again. Let me know if this is also your understanding too.


> They implement a `LanguageInjectionPerformer`

> They implement a `MultiHostInjector`

From [reading the IDEA docs](https://plugins.jetbrains.com/docs/intellij/language-injection.html#languageinjectioncontributor-and-languageinjectionperformer), I believe this is necessary to add 'annotated' permanent injections, like Kotlin has too:

```kotlin

@Language("YAML")
val someConfig = """
hello:
  - I am
  - a language
  - injection
"""
```

I would like this for Nix, but it increeases the complexity a lot like you said, so if I get to it it will be in a separate PR.

> They implement a com.intellij.lang.injection.MultiHostInjector

For this and the latter part of (4), I _think_ that's for injections inside string concatenation too.

As for implementing indent-aware injections in the host, Java escapes its strings keeping track of stateful offsets [here](https://github.com/JetBrains/intellij-community/blob/b6e4bf7c83a723313b263f6dd6124b35f88a5e34/java/java-psi-impl/src/com/intellij/psi/impl/source/javadoc/PsiSnippetDocTagImpl.java#L261), which are populated during `LiteralTextEscaper::decode` [here](https://github.com/JetBrains/intellij-community/blob/b6e4bf7c83a723313b263f6dd6124b35f88a5e34/java/java-psi-impl/src/com/intellij/psi/impl/source/javadoc/PsiSnippetDocTagImpl.java#L235). This bit actually looks simpler in Java, maybe it's because it goes through the single fragment rather than the entire string literal (compare this with [terraform's](https://github.com/JetBrains/intellij-plugins/blob/master/terraform/src/org/intellij/terraform/hcl/psi/impl/HCLStringLiteralTextEscaper.kt#L35)!). I will try to implement this for Nix now see if it works

Cottand commented:
hey @JojOatXGME ! I think this is ready for another review

- I have implemented proper offset tracking and `handleChange`, so now you can launch a code fragment editor and it will get written back to the original file
- I have implemented dealing with the indents inside `decode()`, so the fragment editor will not show the indent from the original Nix file
  - this also means code formatting of the guest language works!

The trade-off is that the `decode()` implementation only works for indented strings for now.

I  have looked at how Java and Kotlin do injections in multiline strings too. They use a different approach (like you said, they work on the fragment directly).  I have yet to understand how they handle string concatenation of each fragment, and how the interpolation plays into that.

Their approach results in a slightly better UX where an interpolation in the host lang (Nix) is not highlighted as the guest lang (YAML, etc) when doing an injection. See the following secreenshots:




Overall I think the former approach, used in this PR, is 'good enough' for me at least. It's also how Terraform does it, even though they also do have string interpolations.

I am still happy to keep looking at the Java/Kotlin approach and come up with better injection in this PR or in a new one.

Feel free to review whenever in your own time. Thank you!

JojOatXGME commented:
Thanks, I will look at it!

> Overall I think the former approach, used in this PR, is 'good enough' for now at least. It's also how Terraform does it, even though they also do have string interpolations.

I am a bit concerned because this means we have two different modes while escaping the text. If a user adds `${...}` in the fragment editor of a bash script, the `${...}` needs to be escaped. However, when the user adds a string interpolation in the Nix file, and then opens the fragment editor, we must not escape the `${...}` from the Nix level interpolation. However, there are probably easy solutions for that. I will take a look at your code with that in mind.

JojOatXGME commented on src/main/java/org/nixos/idea/psi/impl/AbstractNixString.kt:
```diff
@@ -0,0 +1,51 @@
+        // if the first line was removed in the fragment, add it back to preserve a multiline string
+        val withLeadingBlankLine = if (lines.first().isNotEmpty()) listOf("") + withIndent else withIndent
+
+        originalNode?.replaceWithText(withLeadingBlankLine.joinToString(separator = System.lineSeparator()))
```
issue: This line throws an exception when using line feeds which don't match the default on the system. We probably have to use the line feeds configured in IntelliJ for the current file, instead of using `System.lineSeparator()`. Or maybe it works to always use `\n`. Not sure what `replaceWithText` expects here.

<details>
<summary>See exception I got on Windows</summary>

```
java.lang.AssertionError: Wrong line separators: '\r\n  first_...' at offset 0
	at com.intellij.openapi.util.text.StringUtil.assertValidSeparators(StringUtil.java:2484)
	at com.intellij.openapi.editor.impl.DocumentImpl.assertValidSeparators(DocumentImpl.java:706)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:600)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:591)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.doCommitTransaction(PsiToDocumentSynchronizer.java:212)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.lambda$commitTransaction$1(PsiToDocumentSynchronizer.java:188)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.lambda$doSync$0(PsiToDocumentSynchronizer.java:106)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.performAtomically(PsiToDocumentSynchronizer.java:124)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.doSync(PsiToDocumentSynchronizer.java:106)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.commitTransaction(PsiToDocumentSynchronizer.java:188)
	at com.intellij.pom.core.impl.PomModelImpl.commitTransaction(PomModelImpl.java:195)
	at com.intellij.pom.core.impl.PomModelImpl.lambda$runTransaction$1(PomModelImpl.java:151)
	at com.intellij.psi.impl.DebugUtil.performPsiModification(DebugUtil.java:535)
	at com.intellij.pom.core.impl.PomModelImpl.lambda$runTransaction$2(PomModelImpl.java:103)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeNonCancelableSection$2(CoreProgressManager.java:228)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:634)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$computeInNonCancelableSection$3(CoreProgressManager.java:236)
	at com.intellij.openapi.progress.Cancellation.computeInNonCancelableSection(Cancellation.java:53)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeInNonCancelableSection(CoreProgressManager.java:236)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:227)
	at com.intellij.pom.core.impl.PomModelImpl.runTransaction(PomModelImpl.java:92)
	at com.intellij.psi.impl.source.tree.ChangeUtil.prepareAndRunChangeAction(ChangeUtil.java:141)
	at com.intellij.psi.impl.source.tree.CompositeElement.replaceChild(CompositeElement.java:632)
	at com.intellij.psi.impl.source.tree.LeafElement.replaceWithText(LeafElement.java:138)
	at org.nixos.idea.psi.impl.AbstractNixString.updateText(AbstractNixString.kt:43)
	at org.nixos.idea.psi.impl.AbstractNixString.updateText(AbstractNixString.kt:12)
	at org.nixos.idea.psi.NixStringManipulator.handleContentChange(NixStringManipulator.kt:24)
	at org.nixos.idea.psi.NixStringManipulator.handleContentChange(NixStringManipulator.kt:10)
	at com.intellij.psi.ElementManipulators.handleContentChange(ElementManipulators.java:65)
	at com.intellij.psi.impl.source.tree.injected.changesHandler.CommonInjectedFileChangesHandler.updateHostElement(CommonInjectedFileChangesHandler.kt:170)
	at com.intellij.psi.impl.source.tree.injected.changesHandler.CommonInjectedFileChangesHandler.updateHostOrFail(CommonInjectedFileChangesHandler.kt:149)
	at com.intellij.psi.impl.source.tree.injected.changesHandler.CommonInjectedFileChangesHandler.commitToOriginal(CommonInjectedFileChangesHandler.kt:122)
	at com.intellij.codeInsight.intention.impl.QuickEditHandler.lambda$commitToOriginal$10(QuickEditHandler.java:343)
	at com.intellij.psi.impl.source.PostprocessReformattingAspectImpl.lambda$disablePostprocessFormattingInside$1(PostprocessReformattingAspectImpl.java:119)
	at com.intellij.psi.impl.source.PostprocessReformattingAspectImpl.disablePostprocessFormattingInside(PostprocessReformattingAspectImpl.java:128)
	at com.intellij.psi.impl.source.PostprocessReformattingAspectImpl.disablePostprocessFormattingInside(PostprocessReformattingAspectImpl.java:118)
	at com.intellij.codeInsight.intention.impl.QuickEditHandler.commitToOriginal(QuickEditHandler.java:343)
	at com.intellij.codeInsight.intention.impl.QuickEditHandler.documentChanged(QuickEditHandler.java:278)
	at com.intellij.openapi.editor.impl.DocumentImpl.lambda$changedUpdate$1(DocumentImpl.java:913)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeNonCancelableSection$2(CoreProgressManager.java:228)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:685)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:641)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$computeInNonCancelableSection$3(CoreProgressManager.java:236)
	at com.intellij.openapi.progress.Cancellation.computeInNonCancelableSection(Cancellation.java:57)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeInNonCancelableSection(CoreProgressManager.java:236)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:227)
	at com.intellij.openapi.editor.impl.DocumentImpl.changedUpdate(DocumentImpl.java:910)
	at com.intellij.openapi.editor.impl.DocumentImpl.updateText(DocumentImpl.java:814)
	at com.intellij.openapi.editor.impl.DocumentImpl.deleteString(DocumentImpl.java:570)
	at com.intellij.openapi.editor.actions.BackspaceAction.lambda$doBackSpaceAtCaret$0(BackspaceAction.java:65)
	at com.intellij.openapi.editor.ex.util.EditorUtil.runWithAnimationDisabled(EditorUtil.java:1015)
	at com.intellij.openapi.editor.actions.BackspaceAction.doBackSpaceAtCaret(BackspaceAction.java:70)
	at com.intellij.openapi.editor.actions.BackspaceAction$Handler.executeWriteAction(BackspaceAction.java:32)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler$1.run(EditorWriteActionHandler.java:42)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:975)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler.doExecute(EditorWriteActionHandler.java:56)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:202)
	at com.intellij.database.run.SqlNotebookDeleteHandler.doExecute(SqlNotebookDeleteHandler.java:30)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:202)
	at com.intellij.codeInsight.inline.completion.CancellationKeyInlineCompletionHandler.doExecute(InlineCompletionActions.kt:31)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:202)
	at com.intellij.codeInsight.editorActions.BackspaceHandler.handleBackspace(BackspaceHandler.java:91)
	at com.intellij.codeInsight.editorActions.BackspaceHandler.executeWriteAction(BackspaceHandler.java:47)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler$1.run(EditorWriteActionHandler.java:42)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:975)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler.doExecute(EditorWriteActionHandler.java:56)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$2(EditorActionHandler.java:192)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:89)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$3(EditorActionHandler.java:191)
	at com.intellij.openapi.editor.impl.CaretModelImpl.lambda$runForEachCaret$3(CaretModelImpl.java:302)
	at com.intellij.openapi.editor.impl.CaretModelImpl.doWithCaretMerging(CaretModelImpl.java:411)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:311)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:288)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:189)
	at com.intellij.openapi.editor.actions.DeleteSelectionHandler.executeWriteAction(DeleteSelectionHandler.java:33)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler$1.run(EditorWriteActionHandler.java:42)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:975)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler.doExecute(EditorWriteActionHandler.java:56)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$4(EditorActionHandler.java:199)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:89)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:198)
	at com.intellij.codeInsight.lookup.impl.BackspaceHandler.doExecute(BackspaceHandler.java:25)
	at com.intellij.openapi.editor.actionSystem.DynamicEditorActionHandler.doExecute(DynamicEditorActionHandler.java:63)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$4(EditorActionHandler.java:199)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:89)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:198)
	at com.intellij.openapi.editor.actionSystem.EditorAction.lambda$actionPerformed$0(EditorAction.java:92)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:225)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:177)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:101)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:77)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.doPerformActionOrShowPopup(ActionUtil.java:344)
	at com.intellij.openapi.keymap.impl.ActionProcessor.performAction(ActionProcessor.java:32)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$myActionProcessor$1.performAction(IdeKeyEventDispatcher.kt:496)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcherKt.doPerformActionInner$lambda$4$lambda$3(IdeKeyEventDispatcher.kt:831)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:106)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:95)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcherKt.doPerformActionInner$lambda$4(IdeKeyEventDispatcher.kt:831)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:381)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcherKt.doPerformActionInner(IdeKeyEventDispatcher.kt:829)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcherKt.access$doPerformActionInner(IdeKeyEventDispatcher.kt:1)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction$intellij_platform_ide_impl(IdeKeyEventDispatcher.kt:559)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.kt:509)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.kt:448)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.kt:441)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.kt:303)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.kt:620)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$11(IdeEventQueue.kt:581)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithEnabledImplicitRead(RwLockHolder.kt:75)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithImplicitRead(RwLockHolder.kt:67)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:581)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:72)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:355)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:354)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:793)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:354)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:349)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:1014)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:114)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:1014)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:349)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:848)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:391)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

</details>

Cottand replied:
hmmm I am not sure how to get a line separator from the IDEA API. I cannot use `VirtualFile::getDetectedLineSeparator()` so I was guessing from system.

I will replace this with `\n` and hope it resolves your exception :/

JojOatXGME commented on src/main/java/org/nixos/idea/psi/impl/AbstractNixString.kt:
```diff
@@ -0,0 +1,51 @@
+            LOG.info("not a nix ind string")
+            return this
+        }
+        val originalNode = astNode.firstChildNode.treeNext.firstChildNode as? LeafPsiElement
```
issue: This method doesn't seem to escape the input, and may therefore also remove existing escape sequences. For some reason, some lines were also duplicated for me after trying it with the following code:

```nix
pkgs.writeShellScript "my-script.sh" ''
  first_of_array=''${ARRAY[0]}
  from_nix=${lib.escapeShellArg someVar}
  ''
```

Cottand replied:
this is a good catch. I was aware this method did not escape the input and would remove existing escape sequences. This is because at the sage where we get the bash text, I have no way to know if the coming `${` is a bash interpolation or a Nix one, so I don't know whether it should be escaped. Likewise for newlines: If I get a `\n` char, should it get converted to an escaped `''\n` or into an actual `\n`?

I think this is where the method of not injecting in the text fragment may have reached its limits :/

JojOatXGME commented on src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt:
```diff
@@ -0,0 +1,142 @@
+         *
+         * @returns the minIndent of the string if successful, or null if unsuccessful.
+         */
+        fun unescapeAndDecode(chars: String, outChars: StringBuilder, sourceOffsets: IntArray?): Boolean {
```
note: I am not entirely happy with having two implementations to parse/unescape strings now. Anyway, that is not important right now.

JojOatXGME reviewed (changes requested):
Unfortunately, I still found a few issues. The issue of the missing escaping in `AbstractNixString.updateText` is probably tricky to resolve 😟. The remaining topics seem rather trivial (or non-blocking). If you don't mind, I may try out some changes later today, as I got some ideas while looking at your PR.

Cottand commented on src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt:
```diff
@@ -0,0 +1,142 @@
+        val subText: String = rangeInsideHost.substring(myHost.text)
+        val array = IntArray(subText.length + 1)
+        val success = unescapeAndDecode(subText, outChars, array)
+        outSourceOffsets = array
```
I wish I could take credit, I borrowed the idea from terraform

Cottand commented on src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt:
```diff
@@ -0,0 +1,142 @@
+         * Does not consider interpolations so that
+         * they do appear in the guest language and remain when we end up converting back to Nix.
+         *
+         * @returns the minIndent of the string if successful, or null if unsuccessful.
```
good catch, fixed

Cottand commented on src/main/java/org/nixos/idea/psi/impl/AbstractNixString.kt:
```diff
@@ -0,0 +1,51 @@
+abstract class AbstractNixString(private val astNode: ASTNode) : PsiLanguageInjectionHost,
+    AbstractNixPsiElement(astNode), NixString {
+
+    override fun isValidHost() = true
```
yes, good catch. Fixed.

Cottand commented:
Thanks for your review 🙏 

Of course you are welcome to try anything you please!

 As for the escaping issue, you are correct that it's a blocker. I think I might have to re-implement this with injections directly in the text fragment, but I have to learn how it works for Java and Kotlin first. I do not see another way to get around the problem you pointed out. Note Kotlin (which has `${}` interpolatinos) should suffer from the same

JojOatXGME commented:
Just to give you some status update: I noticed that there were also some problems regarding the decoding of the strings. Most noticeably, the indentation is not detected correctly, and I think empty strings would cause issues. (I think the following assumption is not correct.)

https://github.com/NixOS/nix-idea/blob/7d2f79b4dcb042867b783c9e6342e68c9e96ee18/src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt#L75-L77

After my first attempt to detect the indentation correctly (af39db9543c9525b448d167f1e47aaac37f06bf2, only detection of indent), I got the idea that I could adjust the lexer to make the implementation much easier. I have now implemented the correct removal of the indentation at f47a40329bcb6dc4bb3d3d2916a15589869fc025. I also made a first attempt to use this decoding in `NixStringLiteralEscaper` at 328d64404eae6e8b33a79bc5bd5461e4ad7784d3, but it seems unpleasantly complex, and it doesn't work. (The indices I write into `outSourceOffsets` seem incorrect.)

Regarding the main issue with the interpolations, I didn't manage to look into it yet. I may also not have time for this topic for the second half of this week. :(

If you like, I can push these changes on your branch. I wasn't sure if you would like that, as some of my changes are quite different compared to the code you have touched so far, and pushing them into the PR would mean that they would eventually end up in "your" commit after squashing the PR.

Cottand commented:
hey @JojOatXGME ! sorry for the late replies. I will be a bit unresponsive the coming week but I am keeping an eye on this PR.

> If you like, I can push these changes on your branch

Feel free to push into mine, if we need my work we can always branch off at 7d2f79b4dcb042867b783c9e6342e68c9e96ee18 later.

> The indices I write into outSourceOffsets seem incorrect

I had a lot of this in my initially. Over the coming week I will try to look at your new approach and see where the wrong offsets might be!

At any rate thanks for your continued support

JojOatXGME commented:
While looking at this topic again, I noticed the following statement from the Javadoc of `MultiHostInjector`.

> For the more high-level API consider implementing the `LanguageInjectionContributor` and/or `LanguageInjectionPerformer`

I gave it a try and implemented `LanguageInjectionPerformer` at ce8ee544158f7278aad6ee4cb1aef26e0148b4d1. It worked much better than expected. Looks like IDEA implements some magic to make the rest just work. Note that I skipped some parts of what JetBrains' implementations of the interface do, like calling `InjectorUtils.registerSupport`. Not sure for what it is, seems to work without it.

![Screenshot showing injected language with interpolations in host language](https://github.com/user-attachments/assets/b5fa7135-5279-4b7f-9761-2cc7630df54e)

Anyway, while looking into this topic, I also noticed further edge cases and bugs I tried to resolve. This is why the commit is much bigger. One remaining bug is that I get a strange `RuntimeException` when I don't start with an empty line. At some point, I started to look into tests, so that I can put the scenarios into tests before I continue to investigate them. However, I haven't come very far with this yet. See `NixInjectionPerformerTest`. I just pushed what I have right now. I will probably **not** look into this for the next two or three weeks.

I also noticed that JetBrains' documentation at https://plugins.jetbrains.com/docs/intellij/language-injection.html mentions that language plugins should implement `AbstractLanguageInjectionSupport`. I ignored that so far. Doesn't look like we actually need it. Note that to implement it, we would need to add a dependency to the plugin `org.intellij.intelliLang`. (Btw, the same dependency is required to access `InjectorUtils`.)

Cottand commented:
Hey @JojOatXGME I am sorry for the late reply. First off, congrats on your progress with https://github.com/NixOS/nix-idea/pull/82/commits/ce8ee544158f7278aad6ee4cb1aef26e0148b4d1 it does mostly work.

This afternoon I have tried to make sense of the problem you are seeing and implement `NixInjectionPerformerTest` (also unsuccessfully). I am sorry I have not been much use other than with my flakey implementation of the offset-tracking. I would still love to see this feature in Nix-IDEA but I am seriously struggling to understand IDEA's internals

JojOatXGME commented:
No worries. I am also struggling to understand IDEA from time to time. You were already a great help. :+1: While I may have changed some parts of your initial implementation, we wouldn't have brought it that far without your initiative to create some prototype. Unfortunately, I didn't have time over the last two months. Not entirely sure how my next weeks are going to look, but I believe, we can make it work in the not so far future. 😄

FYI, my hypothesis regarding the exception is that IDEA is confused with the first ignored line of the string. The following code in the lexer tells IntelliJ that the first line is insignificant whitespace and can be ignored during parsing.

https://github.com/NixOS/nix-idea/blob/ce8ee544158f7278aad6ee4cb1aef26e0148b4d1/src/main/lang/Nix.flex#L77-L80

My best guess is that some part of IntelliJ is then implicitly truncating this insignificant whitespace during some internal processing, which then causes a different part of IntelliJ to fail, as it didn't expect that the whitespaces are missing now. At least this is the only idea I have right now which could explain the exception. Anyway, I would also still like to have some tests.

Also, if you have any question about the change I did, feel free to ask.

JojOatXGME commented:
> I think the only question I have is around [this test case](https://github.com/NixOS/nix-idea/blob/ce8ee544158f7278aad6ee4cb1aef26e0148b4d1/src/test/java/org/nixos/idea/util/NixStringUtilTest.java#L64) (and similar ones) that fail (as of https://github.com/NixOS/nix-idea/commit/ce8ee544158f7278aad6ee4cb1aef26e0148b4d1). I started looking into fixing them but I do not think I fully understand why you expect it to not have the leading whitespace in the test.

I don't remember if I tried to implement this case yet. I primarily wrote down what I expect as a user in the tests. Adding leading whitespace to an empty line would be unnecessary, as there is nothing to indent. At the same time, this whitespace would also be trailing whitespace. I personally don't like trailing whitespace and try to avoid it. I don't know if my tendency to avoid trailing whitespace is just a personal thing, or a bigger convention. I only know that [Git classifies them as “whitespace errors” by default](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corewhitespace), but that doesn't have to mean much. It is just something Git will highlight in diffs.

FYI: I think you should also be able to create review comments. This way, you can write your comments on the specific lines where you have the questions. Also, if you put your permanent link to a line of code into a separate line, GitHub will actually inline a code snipped showing the code in your comment. (At least if the link points to a line in the same repository, it doesn't work with the link points to a different repository.)
